### PR TITLE
ENH Deprecate clone_from behavior

### PR DESCRIPTION
**Pull request by merveenoyan**
Solves #905.

### CODE REVIEW

**Review by osanseviero — COMMENTED** [review]
Very nice! Thanks for this. Could you also add a small test that checks that a warning is raised please? 

Here is an example: https://github.com/huggingface/huggingface_hub/blob/1d3637c806a1c8067e3bc1255c80069e30e5e889/tests/test_keras_integration.py#L323-L326

**Comment by merveenoyan** [comment]
_plot twist:_ this behavior was used mainly for tests? ✨ so many tests depend on clone_from creating repositories ✨
I took a look at the tests and they have setUp (e.g. `test_clone_with_repo_name_and_org`), which creates a repository remotely so I don't understand why it tries to create a repository with `clone_from` and raises the warning (which is only raised when you get HTTPError for not having the repository remotely). 🦆  I'll take a look tomorrow.

**Comment by merveenoyan** [comment]
I added creation of repositories to setups and managed to reduce failing tests to 3, will look for those on why they fail.

**Comment by merveenoyan** [comment]
maybe @osanseviero knows @LysandreJik 🙂

File: src/huggingface_hub/repository.py [hunk 1]
```diff
@@ -655,6 +656,12 @@ def clone_from(self, repo_url: str, use_auth_token: Union[bool, str, None] = Non
                                 " on Hugging Face Hub."
                             )
                         else:
+                            warnings.warn(
+                                "Creating a repository through clone_from is deprecated"
+                                "will be removed in v0.10.",
```
Comment (osanseviero) [inline src/huggingface_hub/repository.py:661]: Maybe let's change this to v0.11., as 0.9 is already the next release, that way we have 2 releases before deprecating

File: tests/test_repocard.py [hunk 2]
```diff
@@ -204,9 +204,12 @@ def setUpClass(cls):
     def setUp(self) -> None:
         self.repo_path = Path(tempfile.mkdtemp())
         self.REPO_NAME = repo_name()
+        url = self._api.create_repo(
```
Comment (osanseviero) [inline tests/test_repocard.py:207]: When adding some changes, if you have to change tests, that's a red flag to watch out! As you're just adding a warning, nothing should change in any existing tests. It's ok if current tests depend on `clone_from`, I would revert the changes to keep them as they were as we're not changing any behaviour now. We should change the text when actually deprecating this. WDYT?

File: tests/test_repository.py [hunk 3]
```diff
@@ -140,6 +143,19 @@ def test_clone_from_space(self):
                 use_auth_token=self._token,
             )
 
+    def test_clone_from_deprecation_warning(self):
+        """SKIP CREATION"""
```
Comment (osanseviero) [inline tests/test_repository.py:147]: I would remove this docstring as the name of the test is already great :fire:

File: tests/test_repository.py [hunk 4]
```diff
@@ -140,6 +143,19 @@ def test_clone_from_space(self):
...
+        with pytest.raises(
+            FutureWarning,
+            match="Creating a repository through clone_from is deprecated*",
+        ):
+            Repository(
+                WORKING_REPO_DIR,
+                clone_from=f"{USER}/{uuid.uuid4()}",
+                use_auth_token=self._token,
+                repo_type="dataset",
```
Comment (osanseviero) [inline tests/test_repository.py:156]: Nit: feel free to remove `repo_type`. The simpler the test the easier to understand, so anything not strictly needed I would remove

**Review by osanseviero — APPROVED** [review]
Thanks a lot for this PR! I think we can merge after two things:
* I think we should revert the changes in the tests as I would not expect changes in existing tests since we're only adding a warning. That way we're sure nothing changes :) 
* Once the above is reverted, let's get to green and merge :fire:

**Comment by osanseviero** [comment]
@LysandreJik @merveenoyan quick recap of 3rd party libraries with `push_to_hub` that I'm aware of:

* :yellow_circle:  Adapter Transformers (note they pin to be below 0.8) cc @calpt, pointer is at https://github.com/adapter-hub/adapter-transformers/blob/master/src/transformers/utils/hub.py#L1014. So this won't break as they pin the version, but we should update this nevertheless to either create the repo before or use the non-git-based methods
* :green_circle:   ML Agents: they create the repo before cloning it 
* :green_circle: Stable Baselines 3 and RL Zoo: use non-git based methods now! (cc @simoninithomas for visibility)
* :green_circle: Sentence Transformers creates the repo before
* :green_circle: SpaCy uses non-git based methods now! :partying_face:

File: tests/test_repository.py (hunk 4 above)
Comment (merveenoyan) [reply tests/test_repository.py:156]: I did it because there was two of those tests, one was for model repository and one was for dataset repository.

Comment (osanseviero) [reply tests/test_repository.py:156]: I think that given the behaviour you're changing is ok to just cover one case

**Comment by osanseviero** [comment]
Hey @merveenoyan, we will likely do a release next week. There were some very minor changes to be done in the PR before merging, would be great if we could merge the PR some time early *next week.

**Comment by merveenoyan** [comment]
note to self: it's easier to do git revert between two commit hashes instead of having a separate branch and merging on top 😅

**Review by osanseviero — APPROVED** [review]
The PR looks great! Thanks :fire: Let's wait before merging though to check out the failing tests situation

**Comment by SBrandeis** [comment]
Tests are failing because the CI uses `pytest -Werror::FutureWarning` (see https://github.com/huggingface/huggingface_hub/blob/main/.github/workflows/python-tests.yml#L42-L49), introduced in #787 

Linking relevant pytest documentation: https://docs.pytest.org/en/7.1.x/how-to/capture-warnings.html

TL;DR: `FutureWarnings` are turned into exceptions The tests should be updated so that the warning is never raised.

**Comment by osanseviero** [comment]
> TL;DR: FutureWarnings are turned into exceptions The tests should be updated so that the warning is never raised.

Hmm personally I don't like to change the behavior of tests until the actual code is deprecated. Actually, @merveenoyan had already fixed that and I asked her to revert :thinking: :sweat_smile: (sorry Merve) because I found it unexpected to change all the tests based on raising a warning :thinking: so should we revert to that state?

**Comment by osanseviero** [comment]
Given the silence let's do it :) (revert so the tests don't raise warnings)

**Comment by Wauplin** [comment]
If I understood correctly the discussion here, the problem is that when we deprecate some arguments, previous tests are failing because of the `FutureWarnings` been considered as an error. Since changing previously working tests can be annoying as pointed out by @osanseviero, what would you think about a decorator around tests when we expect them to use deprecated arguments ? The general idea is to avoid modifying the test content itself.

A very first draft could be just something like this:

```python
# in tests/testing_utils.py
from functools import wraps
from typing import Callable

def expect_deprecation(test_function:Callable)-> Callable:
    """Decorator to flag tests that we expect to use deprecated arguments.
    
    See discussion in https://github.com/huggingface/huggingface_hub/pull/952.
    """

    @wraps(test_function)
    def _inner_test_function(*args, **kwargs):
        with pytest.warns(FutureWarning, match="*deprecate*"):
            return test_function(*args, **kwargs)

    return _inner_test_function
```

And then:

```python
# in tests/test_something.py
from .testing_utils import expect_deprecation


class SomethingTest(unittest.TestCase):
    (...)

    @expect_deprecation
    def test_that_use_a_deprecated_argument(self):
        (...)
```

This is a completely untested draft to give an idea. For more precise deprecation, we can imagine a `match` argument to `expect_deprecation` so that not all FutureWarnings are caught. I feel that is something we will need in a lot of cases.

@osanseviero @LysandreJik @SBrandeis Any thoughts ?

**Comment by merveenoyan** [comment]
@Wauplin this looks neat, I'll try this now! 🙂

**Comment by Wauplin** [comment]
@merveenoyan FYI, I just implemented the `@expect_deprecation`decorator in another PR (https://github.com/huggingface/huggingface_hub/pull/847) in which I also needed it (exactly same context as here). I also created a decorator `@_deprecate_arguments` to easily deprecate arguments of a function in the same way `@_deprecate_positional_args` is already working. Both decorators now works hand-in-hand (one to deprecate, the other to test). Sorry in advance if it duplicates work you already started 😬

You can have a look here for [@_deprecate_arguments](https://github.com/huggingface/huggingface_hub/pull/847/files#diff-2bc2ffe1f6b4a361c7e049784af91b968ea81ec509e355ffb32f89bde6659f14) and [@expect_deprecation](https://github.com/huggingface/huggingface_hub/pull/847/files#diff-5f63a484ac9f57663f1e8d18ab41afe1d55e7909133191175cbc4943197f39ad) implementations.

**Comment by merveenoyan** [comment]
@Wauplin thanks a lot! Let me know when it's merged so I can use it in this PR.

**Comment by Wauplin** [comment]
@merveenoyan Yup, sorry about that 😬 It was more a "I need it now so let's add it to the PR". Sorry it slows down work here now 😕

**Comment by Wauplin** [comment]
@merveenoyan I'm finally getting back to you !

PR has been merged and you can now use [this decorator](https://github.com/huggingface/huggingface_hub/blob/main/src/huggingface_hub/utils/_deprecation.py#L52) to deprecate arguments and [this one](https://github.com/huggingface/huggingface_hub/blob/main/tests/testing_utils.py#L240) to test it. Hope it will be easy from now on and sorry for the delay ! 😬

**Comment by merveenoyan** [comment]
@Wauplin I feel like it's best if I just close this PR and start from scratch instead of rewinding and rebasing things 😅

**Comment by Wauplin** [comment]
Yes indeed, sorry for the mess

**Comment by merveenoyan** [comment]
@Wauplin  in this case, `clone_from` as an argument is not really deprecated, what is deprecated inside is using `clone_from` to create a new repo from scratch (you should still be able to clone a repo using clone_from argument) in that case, I shouldn’t be decorating the function taking clone_from, but only decorating the tests right?
(however, test decoration goes hand in hand with deprecation so this is not really useful)

**Comment by Wauplin** [comment]
@merveenoyan Indeed we don't have a decorator for deprecating a specific usage of a valid argument (I don't even know how we could do such a decorator 😁 ).

However you can still use the decorator in the tests. Cherry-picking this commit should solve your problem https://github.com/merveenoyan/huggingface_hub/pull/1/commits/19e82fc73145076eeccb5e3f20c34fb4a3438299.

**Comment by merveenoyan** [comment]
@Wauplin as said by Simon, `-W error::FutureWarning` turn matching warnings into exceptions so the future warning is not raised with `pytest.warns` as used by the decorator you've written thus the tests fail. (we need `z ` for errors instead?)
see
```python
def _inner_decorator(test_function: Callable) -> Callable:
        @wraps(test_function)
        def _inner_test_function(*args, **kwargs):
            with pytest.warns(FutureWarning, match=f".*'{function_name}'.*"):
                return test_function(*args, **kwargs)

        return _inner_test_function
```

I will have to edit the tests myself or we can change the `pytest.warns` or how the tests are executed. What do you say?

**Comment by Wauplin** [comment]
In theory `pytest.warns` is catching during the execution of the test. The `-W error::FutureWarning` flag turns the warning into an error only at the end when all tests are done and pytest gather the results.

I think if there is still an issue it less in a typo or something 😕 I'm having a quick look at it.

File: tests/test_repository.py [hunk 5]
```diff
@@ -137,6 +138,17 @@ def test_clone_from_space(self):
                 use_auth_token=self._token,
             )
 
+    def test_clone_from_deprecation_warning(self):
+        with pytest.warns(
+            FutureWarning,
+            match="Creating a repository through clone_from is deprecated*",
```
Comment (Wauplin) [inline tests/test_repository.py:144]:
```suggestion
            match="Creating a repository through 'clone_from' is deprecated*",
```

File: tests/test_repository.py [hunk 6]
```diff
@@ -463,6 +476,7 @@ def test_clone_with_repo_name_and_org(self):
         self.assertTrue("model.bin" in files)
 
     @retry_endpoint
+    @expect_deprecation("clone_from")
```
Comment (Wauplin) [inline tests/test_repository.py:479]:
```suggestion
```

**Comment by Wauplin** [comment]
Seems like:
- `test_clone_with_endpoint` is not decorated.
- `test_clone_with_repo_name_and_user_namespace`: there are 2 of them. They should not be decorated but one is.
- `test_clone_with_repo_name_org_and_no_auth_token `: there are 2 of them. They should be decorated but one is not.

I can't really make the changes directly because it's on a fork 😕

**Review by Wauplin — APPROVED** [review]

**Comment by osanseviero** [comment]
:partying_face:
